File: frameworks/rust/salvo/Implementation/routes/sse.rs
use futures_util::stream::{self, StreamExt};
use salvo::prelude::*;
use salvo::sse::{self, SseEvent};

use crate::Payloads;
use crate::payloads::Payload;

// rb:handler sse.medium
/// items.medium's rows through Salvo's own event stream, which writes each row SseEvent::json
/// serialises as the data of one event.
#[derive(Clone, Copy)]
struct Events(&'static Payload);

#[handler]
impl Events {
    async fn handle(&self, res: &mut Response) {
        sse::stream(res, stream::iter(&self.0.items).map(|row| SseEvent::default().json(row)));
    }
}
// rb:end

/// sse: items.medium's rows as server-sent events.
pub fn router(p: &'static Payloads) -> Router {
    Router::with_path("/sse/medium").get(Events(&p.medium))
}

File: frameworks/python/django-rest-framework/Implementation/views/sse.py
from django.conf import settings
from django.http import StreamingHttpResponse
from rest_framework.decorators import api_view, renderer_classes
from rest_framework.renderers import JSONRenderer

from renderers import EventStreamRenderer

# sse: items.medium's rows as server-sent events. DRF has no event stream of its own, so the view
# frames each row as one event and streams it, as the stream family streams its lines. The request
# accepts only text/event-stream, which DRF's content negotiation refuses with 406 unless a renderer
# of the view offers it, so the view names EventStreamRenderer.
P = settings.PAYLOADS
JSON = JSONRenderer()


# rb:handler sse.medium
@api_view(["GET"])
@renderer_classes([EventStreamRenderer])
def medium(request):
    return StreamingHttpResponse((b"data: " + JSON.render(row) + b"\n\n" for row in P.medium["items"]), content_type="text/event-stream")
